### PR TITLE
Review conditions

### PR DESCRIPTION
**Pull request by EGiataganas**
### Description of change

Adding ability to reviewer for reviewing conditions.
Also introducing review polymorphic table

### Story Link

https://trello.com/c/QLegnd93

### Screenshots

Screenshot 2023-11-02 at 04 51 50

### CODE REVIEW

**Comment by pixeltrix** [comment]
Maybe extract the `Review` -> `Reviewing` change into a separate PR and get that merged?

**Comment by EGiataganas** [comment]
@pixeltrix I was thinking the same thing! That's why I've made it in a separate commit. Pushing now new branch and PR #1344

**Comment by EGiataganas** [comment]
Adding #1351 to be able to review overall conditions

File: app/controllers/planning_applications/review/conditions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        respond_to do |format|
+          format.html
+        end
+      end
+
+      def edit
+        respond_to do |format|
+          format.html
+        end
+      end
```
Comment (pixeltrix) [inline app/controllers/planning_applications/review/conditions_controller.rb:25]: The `show` and `editable` views are the same apart from `editable` being passed to the `form` partial being `true/false` - we should only have the `show` action and use the status to determine the value of `editable`.

File: app/controllers/planning_applications/review/conditions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+      def return_to_officer?
+        params.dig(:condition_set, :action) == "rejected"
+      end
+
+      def ensure_user_is_reviewer
+        current_user.reviewer?
+      end
```
Comment (pixeltrix) [inline app/controllers/planning_applications/review/conditions_controller.rb:109]: This appears to be all for setting the status on the `ConditionSet` depending on what the review is - in which case it should move to callbacks on the `Reviewable` concern and this controller should only be updating the review directly.

File: app/models/concerns/reviewable.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+module Reviewable
+  extend ActiveSupport::Concern
+
+  included do
+    has_one :review, as: :reviewable, dependent: :destroy
```
Comment (pixeltrix) [inline app/models/concerns/reviewable.rb:7]: Not for this PR but what should happen if there are multiple rounds of reviews where the conditions are sent back for amendments by the reviewer? We should probably track that history (especially for minors/majors) so we might need to change this to a `has_many` and have a setup where we only append new reviews and show the latest.

**Review by pixeltrix — CHANGES_REQUESTED** [review]

File: app/models/condition.rb [hunk 4]
```diff
@@ -13,6 +13,10 @@ def sort_key
     [title_key, timestamp_key, Float::INFINITY].compact.first
   end
 
+  def review_title
+    title.blank? ? "Other" : title
```
Comment (benjamineskola) [inline app/models/condition.rb:17]:
```suggestion
    title.presence || "Other"
```

> `object.presence` is equivalent to `object.present? ? object : nil`
[\[docs\]](https://api.rubyonrails.org/classes/Object.html#method-i-presence)

File: app/controllers/planning_applications/review/tasks_controller.rb [hunk 5]
```diff
@@ -5,12 +5,24 @@ module Review
...
+
+      private
+
+      def set_condition_set
+        @condition_set = @planning_application.condition_set || @planning_application.create_condition_set!
+      end
+
+      def set_condition_set_review
+        @condition_set_review = @condition_set.review || @condition_set.create_review!
+      end
```
[the remark this answers is not in the document]
Comment (benjamineskola) [reply app/controllers/planning_applications/review/tasks_controller.rb:25]: Yes, I think it would be good to avoid a situation like we had with consultations, where it was possible for the object not to have been created when already trying to display the status of it.

File: app/models/concerns/reviewable.rb (hunk 3 above)
Comment (benbaumann95) [reply app/models/concerns/reviewable.rb:7]: Agreed, needs to be a `has_many` reviews as there can be multiple rounds

File: app/components/task_list_items/reviewing/condition_component.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      def status_tag_component
+        StatusTags::BaseComponent.new(status:)
+      end
+
+      def status
+        if condition_set.review.present?
+          condition_set.review.status
+        else
+          :in_progress
+        end
```
[the remark this answers is not in the document]
Comment (EGiataganas) [reply app/components/task_list_items/reviewing/condition_component.rb:36]: Double checked it and no need for this condition, will leave `condition_set.review.status`

File: app/models/concerns/reviewable.rb (hunk 3 above)
Comment (EGiataganas) [reply app/models/concerns/reviewable.rb:7]: Agree, but as I didn't want to expand it more I've left it as `has_one :review` which is what is needed for this

File: app/controllers/planning_applications/review/tasks_controller.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (benjamineskola) [reply app/controllers/planning_applications/review/tasks_controller.rb:25]: Yes, I think it would be good to avoid a situation like we had with consultations, where it was possible for the object not to have been created when already trying to display the status of it.

(comment moved to correct place 🤦🏻‍♂️)

File: app/controllers/planning_applications/review/conditions_controller.rb [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+      def condition_set_review_params
+        params.require(:condition_set)
+          .permit(review_attributes: %i[action comment],
+            conditions_attributes: %i[_destroy id standard title text reason])
+          .to_h
+          .deep_merge(
+            status: condition_set_status,
+            review_attributes: {reviewed_at: Time.current, reviewer: current_user, status: status}
+          )
```
Comment (pixeltrix) [inline app/controllers/planning_applications/review/conditions_controller.rb:47]: These really should be set in the model via callbacks or arguments to a custom method but I think we agreed to leave it to a separate PR, right?

**Review by pixeltrix — APPROVED** [review]
Looks good 👍🏻

File: app/controllers/planning_applications/review/conditions_controller.rb (hunk 7 above)
Comment (EGiataganas) [reply app/controllers/planning_applications/review/conditions_controller.rb:47]: Yes, I agree! We will make a follow up pr to also replace `CommitMatchable`
